### PR TITLE
Create generic service

### PR DESCRIPTION
**Pull request by thanhtrantruong47**
Example get product by category:

Response : 
image

MockApi: https://6684bd0356e7503d1ae0ff6f.mockapi.io/products

### CODE REVIEW

File: src/services/Service.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+   * @param {string} id - The ID of the item to fetch.
+   * @returns {Promise<T>} A promise that resolves to an item of type T.
+   */
+  async getById(id: string): Promise<T> {
+    try {
+      const response = await fetch(`${this.resourceUrl}/${id}`);
+      if (!response.ok) {
+        throw new Error('Failed to fetch data');
+      }
+      return await response.json();
```
Comment (thanhnguyen-agilityio) [inline src/services/Service.ts]: Is `await` redundant here?

File: src/services/ProductService.ts [hunk 2]
```diff
@@ -1,54 +0,0 @@
...
-  constructor() {
-    super('products');
-  }
-
-  /**
-   * Fetches a product by its ID from the server.
-   * @param {string} id - The ID of the product to fetch.
-   * @returns {Promise<Product>} A promise that resolves to a Product object.
-   */
-  async getById(id: string): Promise<Product> {
```
Comment (thanhnguyen-agilityio) [inline src/services/ProductService.ts]: Is this duplicate with base service code?

File: src/types/product.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+export interface Product {
```
Comment (thanhnguyen-agilityio) [inline src/types/product.ts:1]: Recheck which fields could be optional

File: src/services/HttpService.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+      throw new Error(`Failed to create item: ${errorMessage}`);
+    }
+  }
+
+  /**
+   * Deletes an item by its ID from the server.
+   * @param {string} id - The ID of the item to delete.
+   * @returns {Promise<void>} A promise that resolves when the item is successfully deleted.
+   */
+  async deleteById(id: string): Promise<void> {
```
Comment (thanhnguyen-agilityio) [inline src/services/HttpService.ts]: `delete`

File: src/services/HttpService.ts [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    }
+  }
+
+  /**
+   * Updates an item by its ID on the server.
+   * @param {string} id - The ID of the item to update.
+   * @param {T} data - The item data to update.
+   * @returns {Promise<T>} A promise that resolves to the updated item of type T.
+   */
+  async updateById(id: string, data: T): Promise<T> {
```
Comment (thanhnguyen-agilityio) [inline src/services/HttpService.ts]: `update`

File: src/types/product.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+  id: number;
+  title: string;
+  price: number;
+  images: string[];
+  colors?: string[];
+  sizes?: string[];
+  description: string;
+  stock: number;
+  rate?: number;
```
Comment (thanhnguyen-agilityio) [inline src/types/product.ts:10]: What diff between `rate` and `numberRating`?

File: src/services/APIService.ts [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+  /**
+   * Fetches an item by its ID from the server using a GET request.
+   * @param {string} id - The ID of the item to fetch.
+   * @returns {Promise<Response<T>>} A promise that resolves to a Response object containing the fetched item.
+   */
+  async getById(id: string): Promise<Response<T>> {
+    try {
+      const response = await fetch(`${this.resourceUrl}/${id}`, {
+        method: 'GET',
+        headers: { 'Content-Type': 'application/json' },
```
Comment (hungdo-agilityio) [inline src/services/APIService.ts]: Do we need headers for GET method

File: src/App.tsx [hunk 8]
```diff
@@ -1,7 +1,24 @@
...
+      const response = await productAPIService.getById('100');
+      if (!response.data) {
+        console.log('No data');
+      } else {
+        console.log('Products:', response.data);
+      }
+    };
+
+    fetchProducts();
+  }, []);
```
Comment (hungdo-agilityio) [inline src/App.tsx]: Please add note/comments for demo code

File: src/services/APIService.ts [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+  async delete(id: string): Promise<Response<void>> {
+    try {
+      const response = await fetch(`${this.resourceUrl}/${id}`, {
+        method: 'DELETE',
+        headers: { 'Content-Type': 'application/json' },
+      });
+      if (!response.ok) {
+        throw new Error('Failed to delete item');
+      }
+      return {};
```
Comment (hungdo-agilityio) [inline src/services/APIService.ts]: Why return here

File: src/types/response.ts [hunk 10]
```diff
@@ -0,0 +1,4 @@
+export interface Response<T> {
```
Comment (hungdo-agilityio) [inline src/types/response.ts]: `ApiResponse` instead we may have multiple type of Response. Also the file name should be `api.ts`

File: src/services/APIService.ts [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        throw new Error('Failed to fetch data');
+      }
+      const data = await response.json();
+      return { data };
+    } catch (error: unknown) {
+      const errorMessage =
+        error instanceof Error ? error.message : 'Unknown error';
+      console.error('Error fetching list', error);
+      return {
+        error: new Error(`Error occurred during item search: ${errorMessage}`),
```
Comment (hungdo-agilityio) [inline src/services/APIService.ts]: Recheck the message

File: src/services/APIService.ts [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+    }
+  }
+
+  /**
+   * Fetches items by category from the server.
+   * @param {string} category - The category of the items to fetch.
+   * @returns {Promise<Response<T[]>>} A promise that resolves to a Response object containing an array of items.
+   */
+  async getList(category?: string): Promise<Response<T[]>> {
+    const url = new URL(this.resourceUrl);
```
Comment (hungdo-agilityio) [inline src/services/APIService.ts:41]: Why args is specific to category

**Review by hungdo-agilityio — APPROVED** [review]

File: src/App.tsx [hunk 13]
```diff
@@ -1,7 +1,28 @@
...
 const App = () => {
+  // example start
+  const productAPIService = new ProductAPIService();
+
+  useEffect(() => {
+    const fetchProducts = async () => {
+      const params = { category: '3', title: 'title 3' };
+      const response = await productAPIService.getList(params);
+
+      if (response.error) {
```
Comment (thanhnguyen-agilityio) [inline src/App.tsx]: `!response.isSuccess`

File: src/services/APIService.ts [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+   * @param {Record<string, string>} params - The query parameters.
+   * @returns {Promise<ApiResponse<T[]>>} A promise that resolves to a Response object containing an array of items.
+   */
+  async getList(params?: Record<string, string>): Promise<ApiResponse<T[]>> {
+    const url = new URL(this.resourceUrl);
+    if (params) {
+      Object.keys(params).forEach(key => {
+        url.searchParams.append(key, params[key]);
+      });
+    }
```
Comment (thanhnguyen-agilityio) [inline src/services/APIService.ts:48]: one line break

File: src/services/APIService.ts [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+  }
+
+  /**
+   * Fetches items from the server based on the provided query parameters.
+   * @param {Record<string, string>} params - The query parameters.
+   * @returns {Promise<ApiResponse<T[]>>} A promise that resolves to a Response object containing an array of items.
+   */
+  async getList(params?: Record<string, string>): Promise<ApiResponse<T[]>> {
+    const url = new URL(this.resourceUrl);
+    if (params) {
```
Comment (thanhnguyen-agilityio) [inline src/services/APIService.ts:44]: comment code: Handle query param with key:value

File: src/services/APIService.ts [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+      Object.keys(params).forEach(key => {
+        url.searchParams.append(key, params[key]);
+      });
+    }
+    try {
+      const response = await fetch(url.toString());
+      const data = await response.json();
+
+      if (!response.ok) {
+        throw new Error('Failed to fetch data');
```
Comment (thanhnguyen-agilityio) [inline src/services/APIService.ts]: Not correct response of error as discuss, not throw error anymore.
should be:
```
return {
  isSuccess: false,
  errors: [new Error('Failed to fetch data')]
}
```

File: src/services/APIService.ts [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+  /**
+   * Helper function to handle errors consistently.
+   * @param {unknown} error - The error object.
+   * @param {string} message - The error message.
+   * @returns {ApiResponse} An object containing the error details.
+   */
+  handleError(error: unknown, message: string) {
+    const errorMessage =
+      error instanceof Error ? error.message : 'Unknown error';
```
Comment (thanhnguyen-agilityio) [inline src/services/APIService.ts]: Is there a case when error is not `Error`, or you guess it?

File: src/types/apiResponse .ts [hunk 18]
```diff
@@ -0,0 +1,4 @@
+export interface ApiResponse<T> {
+  data?: T;
```
Comment (thanhnguyen-agilityio) [inline src/types/apiResponse .ts:2]: Missing `isSuccess` flag. which is boolean.

**Review by thanhnguyen-agilityio — APPROVED** [review]

**Pull request merged**